pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java

```diff
@@ -677,14 +678,30 @@ public RebalanceResult rebalance(
         if (dryRunResult.getStatus() == RebalanceResult.Status.DONE) {
           // If dry-run succeeded, run rebalance asynchronously
           rebalanceConfig.setDryRun(false);
-          _executorService.submit(() -> {
+          Future<RebalanceResult> rebalanceResultFuture = _executorService.submit(() -> {
             try {
-              _pinotHelixResourceManager.rebalanceTable(tableNameWithType, rebalanceConfig, rebalanceJobId, true);
+              return _pinotHelixResourceManager.rebalanceTable(
+                  tableNameWithType, rebalanceConfig, rebalanceJobId, true);
             } catch (Throwable t) {
-              LOGGER.error("Caught exception/error while rebalancing table: {}", tableNameWithType, t);
+              String errorMsg = String.format("Caught exception/error while rebalancing table: %s", tableNameWithType);
+              LOGGER.error(errorMsg, t);
+              return new RebalanceResult(rebalanceJobId, RebalanceResult.Status.FAILED, errorMsg, null, null, null);
             }
           });
-          waitForJobIdToPersist(dryRunResult.getJobId(), tableNameWithType);
+          boolean isJobIdPersisted = waitForJobIdToPersist(dryRunResult.getJobId(), tableNameWithType);
+
+          // It's possible the dryRun indicates a rebalance is needed, but next rebalance is not.
+          // In that case, we never persist the jobId, but will already have a rebalanceResult.
+          // Return the actual rebalanceResult in this case.
+          if (!isJobIdPersisted && rebalanceResultFuture.isDone()) {
```

Comment:
I did this mainly to try to preserve behavior. But we could also only check `rebalanceResultFuture.isDone()`